primitives/treasury-funding/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+/// * x is the system utilization rate, i.e. fraction of the total storage being provided in the system
+/// that is currently being used.
+/// * d is the falloff or `decay_rate`
+/// * x_ideal: the ideal system utilization rate.
+///
+/// The result is meant to be scaled with the minimum percentage amount and maximum percentage amount
+///  that goes to the treasury.
+///
+/// Arguments are:
+/// * `system_utilization`: The fraction of the total storage being provided in the system that is
+/// currently being used. Known as `x` in the literature. Must be between 0 and 1.
+/// * `ideal_system_utilization`: The fraction of total storage being provided in the system that should
+/// be being actively used. Known as `x_ideal` in the literature. Must be between 0 and 1.
+/// * `falloff`: Known as `decay_rate` in the literature. A co-efficient dictating the strength of
+///   the global incentivization to get the `ideal_system_utilization`. A higher number results in less typical
+///   funds to the treasury at the cost of greater volatility for providers. Must be more than 0.01.
+pub fn compute_percentage_to_treasury<P: PerThing>(
+    system_utilization: P,
+    ideal_system_utilization: P,
+    falloff: P,
+) -> P {
+    if system_utilization < ideal_system_utilization {
+        // ideal_system_utilization is more than 0 because it is strictly more than system_utilization
+        return (system_utilization / ideal_system_utilization).left_from_one();
+    }
+
+    if falloff < P::from_percent(1.into()) {
+        log::error!("Invalid inflation computation: falloff less than 1% is not supported");
+        return PerThing::zero();
+    }
+
+    let accuracy = {
+        let mut a = BigUint::from(Into::<u128>::into(P::ACCURACY));
+        a.lstrip();
+        a
+    };
+
+    let mut falloff = BigUint::from(falloff.deconstruct().into());
+    falloff.lstrip();
+
+    let ln2 = {
+        /// `ln(2)` expressed in as perquintillionth.
+        const LN2: u64 = 0_693_147_180_559_945_309;
+        let ln2 = P::from_rational(LN2.into(), Perquintill::ACCURACY.into());
+        BigUint::from(ln2.deconstruct().into())
+    };
+
+    // falloff is stripped above.
+    let ln2_div_d = div_by_stripped(ln2.mul(&accuracy), &falloff);
+
+    let ftt_param = FTTParam {
+        x_ideal: BigUint::from(ideal_system_utilization.deconstruct().into()),
+        x: BigUint::from(system_utilization.deconstruct().into()),
+        accuracy,
+        ln2_div_d,
+    };
+
+    let res = compute_taylor_serie_part(&ftt_param);
+
+    match u128::try_from(res.clone()) {
+        Ok(res) if res <= Into::<u128>::into(P::ACCURACY) => {
+            P::from_parts(res.saturated_into()).left_from_one()
+        }
+        // If result is beyond bounds there is nothing we can do
+        _ => {
+            log::error!(
+                "Invalid funds to treasury computation: unexpected result {:?}",
+                res
+            );
+            P::zero()
+        }
+    }
+}
```

Comment:
Why do you need a taylor series approximation for a cuadratic function?

Reply:
It's not cuadratic! It's an exponential function with base 2 (basically 2^x), so we use a taylor series to make it more computationally efficient while losing little precision.